api/service.go

```diff
@@ -693,31 +736,37 @@ func (s *service) AddFile(server pb.API_AddFileServer) error {
 	if err != nil {
 		return err
 	}
-	var filePath string
+	var projectID, filePath string
 	switch payload := req.Payload.(type) {
-	case *pb.AddFileRequest_Header_:
+	case *pb.PushBucketPathRequest_Header_:
+		projectID = payload.Header.ProjectID
 		filePath = payload.Header.Path
 	default:
 		return fmt.Errorf("add file header is required")
 	}
-	folderName, fileName := parsePath(filePath)
-	folder, err := s.getFolderWithScope(server.Context(), folderName, scope)
+	buckName, fileName := parsePath(filePath)
+	buck, err := s.getBucketWithScope(server.Context(), buckName, scope)
 	if err != nil {
-		return err
+		if status.Convert(err).Code() != codes.NotFound {
+			return err
+		}
+		buck, err = s.createBucket(server.Context(), buckName, false, projectID)
```

Comment:
Yep. It's not totally clear to me how to deal with "public" yet. I think a public bucket just means it's viewable at a static path over the gateway. However, some addition thoughts here / a proposal: 

It would be nice to be able to mark _any_ path as public, much like S3. This might involve an additional badger key space in `textile` that just tracks a path (as key) and a boolean public value. So, everything is implied to be private, sans for the keys given in the "public" keyspace. When traversing files/dirs in the gateway, a fast check would be made to ensure a path or a parent of the path has a public entry.

Reply:
Sounds good.
Today I was also thinking about if the `.seed` the idea can be extended further and enable folder-configuration _embedded_ in the filesystem. It may serve for permissions, or even flagging it _public_ or _private_ too. 

So, someone can change the visibility just by editing the folder configuration file?
Maybe that's easier for people to change things since is just editing the file and pushing changes. But that would add some work to: validate changes, check format, etc.

Just brainstorming too.

Reply:
Nice, that does sound like a more extensible solution. Quite like S3 bucket policy files: https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html

Reply:
Created https://github.com/textileio/textile/issues/84 to continue convo / work.